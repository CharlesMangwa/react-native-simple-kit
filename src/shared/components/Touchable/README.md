# Touchable 👇

A generic component for handling touchable items with specific platforms behaviours on iOS & Android.

### API

- children?: `React$Element<*>`
- onLongPress?: `Function`
- onPress?: `Function`
- ripple?: `{ color: string, borderless: boolean }`
- style?: `StyleSheet`

### Example

```js
/* @flow */
import React from 'react'
import { Alert, StyleSheet, Text } from 'react-native'
import Touchable from '@components/Touchable'

const MyComponent = (): React$Element<*> => (
  <Touchable
    onLongPress={(): void => Alert.alert('👋')}
    onPress={(): void => Alert.alert('🤝')}
    ripple={{ color: #FF990080, borderless: true }}
    style={StyleSheet.flatten({ backgroundColor: 'pink' })}
  >
    <Text>Hello my dear!</Text>
  </Touchable>
)

export default MyComponent
```

## Changelog

#### Thursday, August 2nd

- Implement ripple effect support.

#### Saturday, June 23rd

- Refactoring for React 16.3+

#### Wednesday, November 22th

- Switch to `React.PureComponent`.

#### Thursday, November 16th

- Initial release. We didn't track changes before this version.
