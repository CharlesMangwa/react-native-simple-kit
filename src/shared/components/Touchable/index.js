/* @flow */
/* eslint camelcase: 0 */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  Platform,
  StyleSheet,
  TouchableNativeFeedback,
  TouchableOpacity,
  View,
} from 'react-native'

type DefaultProps = {
  children?: React$Element<*> | Array<React$Element<*>>,
  onLongPress?: Function,
  onPress?: Function,
  ripple?: { color: string, borderless: boolean },
  style?: StyleSheet,
}

type Props = {
  children?: React$Element<*> | Array<React$Element<*>>,
  onLongPress?: Function,
  onPress: Function,
  ripple?: { color: string, borderless: boolean },
  style?: StyleSheet,
}

class Touchable extends Component<Props> {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.arrayOf(PropTypes.element),
    ]),
    onLongPress: PropTypes.func,
    onPress: PropTypes.func.isRequired,
    ripple: PropTypes.shape({
      color: PropTypes.string.isRequired,
      borderless: PropTypes.bool.isRequired,
    }),
    style: PropTypes.any, // eslint-disable-line
  }

  static defaultProps: DefaultProps = {
    children: undefined,
    onLongPress: undefined,
    ripple: undefined,
    style: null,
  }

  UNSAFE_componentWillMount() {
    const { children } = this.props
    if (!children) throw new Error('Touchable requires at least 1 children')
  }

  render() {
    const {
      children,
      onLongPress,
      onPress,
      ripple,
      style,
      ...remainingProps
    } = this.props
    return Platform.OS === 'ios' ? (
      <TouchableOpacity
        activeOpacity={0.7}
        delayLongPress={200}
        onLongPress={onLongPress}
        onPress={onPress}
      >
        <View {...remainingProps} style={style}>
          {children}
        </View>
      </TouchableOpacity>
    ) : (
      <TouchableNativeFeedback
        useForeground
        delayLongPress={200}
        onPress={onPress}
        onLongPress={onLongPress}
        background={
          ripple
            ? TouchableNativeFeedback.Ripple(ripple.color, ripple.borderless)
            : TouchableNativeFeedback.SelectableBackground()
        }
      >
        <View {...remainingProps} style={style}>
          {children}
        </View>
      </TouchableNativeFeedback>
    )
  }
}

export default Touchable
