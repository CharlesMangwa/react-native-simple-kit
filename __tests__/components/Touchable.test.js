/* eslint-disable */

import React from 'react'
import TestRenderer from 'react-test-renderer'
import ShallowRenderer from 'react-test-renderer/shallow'
import { Text } from 'react-native'
import Touchable from '@components/Touchable'

describe('Generic component: <Touchable />', () => {
  let component, onLongPress, onPress, output, renderer

  beforeEach(() => {
    onPress = jest.fn()
    onLongPress = jest.fn()
    renderer = new ShallowRenderer()

    component = TestRenderer.create(
      <Touchable>
        <Text>Hey!</Text>
      </Touchable>
    )

    renderer.render(
      <Touchable onLongPress={() => onLongPress()} onPress={() => onPress()}>
        <Text>Doing well?</Text>
      </Touchable>
    )
    output = renderer.getRenderOutput()
  })

  afterEach(() => jest.clearAllMocks())

  it('renders correctly', () => {
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls onPress when pressed', () => {
    output.props.onPress()
    expect(onPress).toBeCalled()
  })

  it('calls onLongPress when pressed', () => {
    output.props.onLongPress()
    expect(onLongPress).toBeCalled()
  })

  it('throws an error when no children is passed', () => {
    const renderer = new ShallowRenderer()
    expect(() => renderer.render(<Touchable />)).toThrowError(
      'Touchable requires at least 1 children'
    )
  })
})
