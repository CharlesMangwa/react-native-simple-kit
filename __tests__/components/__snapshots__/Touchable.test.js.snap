// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic component: <Touchable /> renders correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={null}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Hey!
    </Text>
  </View>
</View>
`;
